internal/reliable_topic.go

```diff
@@ -0,0 +1,284 @@
+	case core.TopicOverLoadPolicyDiscardOldest:
+		err = r.addOrOverwrite(reliableMsg)
+	}
+	return err
+}
+
+func (r *ReliableTopicProxy) addOrFail(message *reliabletopic.Message) (err error) {
+	seqID, err := r.ringBuffer.Add(message, core.OverflowPolicyFail)
+	if err != nil {
+		return err
+	}
+	if seqID == -1 {
+		// TODO :: add message to error string
+		errorMsg := "failed to publish message to topic: " + r.name
+		return core.NewHazelcastTopicOverflowError(errorMsg, nil)
+	}
+	return
+}
+
+func (r *ReliableTopicProxy) addWithBackoff(message *reliabletopic.Message) (err error) {
+	sleepTime := initialBackoff
+	for {
+		result, err := r.ringBuffer.Add(message, core.OverflowPolicyFail)
+		if err != nil {
+			return err
+		}
+		if result != -1 {
+			return nil
+		}
+		time.Sleep(sleepTime)
+		sleepTime *= 2
+		if sleepTime > maxBackoff {
+			sleepTime = maxBackoff
+		}
+	}
+}
+
+func (r *ReliableTopicProxy) addOrOverwrite(message *reliabletopic.Message) (err error) {
+	_, err = r.ringBuffer.Add(message, core.OverflowPolicyOverwrite)
+	return
+}
+func (r *ReliableTopicProxy) Ringbuffer() core.Ringbuffer {
+	return r.ringBuffer
+}
+
+type messageProcessor struct {
+	id        string
+	sequence  int64
+	cancelled atomic.Value
+	listener  core.ReliableMessageListener
+	proxy     *ReliableTopicProxy
+}
+
+func newMessageProcessor(id string, listener core.ReliableMessageListener, proxy *ReliableTopicProxy) *messageProcessor {
+	msgProcessor := &messageProcessor{
+		id:       id,
+		listener: listener,
+		proxy:    proxy,
+	}
+	msgProcessor.cancelled.Store(false)
+	initialSeq := listener.RetrieveInitialSequence()
+	if initialSeq == -1 {
+		// set initial seq as tail sequence + 1 so that we wont be reading the items that were added before the listener
+		// is added.
+		tailSeq, _ := msgProcessor.proxy.ringBuffer.TailSequence()
+		initialSeq = tailSeq + 1
+	}
+	msgProcessor.sequence = initialSeq
+	return msgProcessor
+}
+
+func (m *messageProcessor) next() {
+	if m.cancelled.Load() == true {
+		return
+	}
+	readResults, err := m.proxy.ringBuffer.ReadMany(m.sequence, 1,
+		m.proxy.config.ReadBatchSize(), nil)
+	if err == nil {
+		m.onResponse(readResults)
+	} else {
+		m.onFailure(err)
+	}
+}
+
+func (m *messageProcessor) onFailure(err error) {
+	if m.cancelled.Load() == true {
+		return
+	}
+	if strings.Contains(err.Error(), "com.hazelcast.ringbuffer.StaleSequenceException") {
+		if m.listener.IsLossTolerant() {
+			headSeq, _ := m.proxy.ringBuffer.HeadSequence()
+			msg := "Topic " + m.proxy.name + " ran into a stale sequence. Jumping from old sequence " +
+				strconv.Itoa(int(m.sequence)) + " " +
+				" to new sequence " + strconv.Itoa(int(headSeq))
+			log.Println(msg)
+			m.sequence = headSeq
+			go m.next()
+			return
+		}
+
+	}
+	msg := "Terminating Message Listener: " + m.id + " on topic: " + m.proxy.name + ". Reason: " + err.Error()
+	log.Println(msg)
+	m.proxy.msgProcessorsMu.Lock()
+	m.cancel()
+	m.proxy.msgProcessorsMu.Unlock()
+}
```

Comment:
Missing handling of other exceptions handled at Java such as HazelcastInstanceNotActiveException, DistributedObjectDestroyedException, HazelcastClientNotActiveException || t instanceof RejectedExecutionException.